lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1706,116 +1713,362 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      }
 
       // If self does not have reference semantics, it is very unlikely that
       // capturing it will create a reference cycle.
-      if (!ty->hasReferenceSemantics())
-        return false;
+      if (!captureType->hasReferenceSemantics()) {
+        requiresSelfQualification = false;
+      }
 
       if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
-        if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
-          // If this capture is using the name `self` actually referring
-          // to some other variable (e.g. with `[self = "hello"]`)
-          // then implicit self is not allowed.
-          if (!selfDecl->isSelfParamCapture()) {
-            return true;
+        auto capturedSelfDecl = closureExpr->getCapturedSelfDecl();
+
+        // If this closure doesn't capture self explicitly, but this closure
+        // requires self qualification, then implicit self is disallowed.
+        if (!capturedSelfDecl && requiresSelfQualification) {
+          return false;
+        }
+
+        // If the closure has an explicit capture using the name `self` that
+        // actually refers to some other variable (e.g. `[self = "hello"]`)
+        // then implicit self is not allowed.
+        if (capturedSelfDecl && !isSimpleSelfCapture(capturedSelfDecl)) {
+          return false;
+        }
+      }
+
+      // If the self decl comes from a conditional statement, validate
+      // that it is an allowed `guard let self` or `if let self` condition.
+      //  - Even if this closure doesn't have a `weak self` capture, it could
+      //    be a closure nested in some parent closure with a `weak self`
+      //    capture, so we should always validate the conditional statement
+      //    that defines self if present.
+      if (validateSelfRebindings) {
+        if (auto conditionalStmt = parentConditionalStmt(selfDecl)) {
+          if (!hasValidSelfRebinding(conditionalStmt, inClosure)) {
+            return false;
           }
         }
       }
 
-      if (auto var = dyn_cast<VarDecl>(DRE->getDecl())) {
-        if (!isEnclosingSelfReference(var, inClosure)) {
+      // If this closure has a `weak self` capture, require that the
+      // closure unwraps self. If not, implicit self is not allowed
+      // in this closure or in any nested closure.
+      if (closureHasWeakSelfCapture(inClosure) &&
+          !hasValidSelfRebinding(parentConditionalStmt(selfDecl), inClosure)) {
+        return false;
+      }
+
+      if (auto autoclosure = dyn_cast<AutoClosureExpr>(inClosure)) {
+        // Implicit self is always allowed in autoclosure thunks generated
+        // during type checking. An example of this is when storing an instance
+        // method as a closure (e.g. `let closure = someInstanceMethodOnSelf`).
+        auto isThunk =
+            autoclosure->getThunkKind() != AutoClosureExpr::Kind::None;
+        if (isThunk) {
+          return true;
+        }
```

Comment:
Sorry I didn't catch this before (and sorry to keep piling on more counterexamples 😅), but I just noticed that this would allow the following to compile, which is currently not allowed in 5.10 (and seems like it should remain illegal):

```swift
class C {
  func bar() -> Int { 0 }
  func foo() async {
    let _ = {
      async let x = bar()
    }
  }
}
```

I think we still need to check parent closures in this case.

Reply:
Good catch, thanks! Fixed.